---
title: "db_inquiry"
format: html
editor: source
---


```{r}
if (!require(librarian)){
  install.packages("librarian")
  library(librarian)
}

# librarian downloads, if not already downloaded, and reads in needed packages
librarian::shelf(tidyverse, dbplyr, here, janitor, lubridate, RPostgres, stringr, DBI, parsedate, uuid, hms, RIBBiTR-BII/ribbitrrr)

# connect to database
dbcon = hopToDB("ribbitr")

```

## Pointers to relevant tables

```{r}
# pull relevant chain tables from DB
db_sample = tbl(dbcon, Id("survey_data", "sample"))
db_bd = tbl(dbcon, Id("survey_data", "bd_qpcr_results"))

db_aural = tbl(dbcon, Id("survey_data", "aural"))
db_capture = tbl(dbcon, Id("survey_data", "capture"))
db_edna = tbl(dbcon, Id("survey_data", "edna"))
db_env = tbl(dbcon, Id("survey_data", "environmental"))
db_ves = tbl(dbcon, Id("survey_data", "ves"))

db_survey = tbl(dbcon, Id("survey_data", "survey"))
db_visit = tbl(dbcon, Id("survey_data", "visit"))
db_site = tbl(dbcon, Id("survey_data", "site"))
db_region = tbl(dbcon, Id("survey_data", "region"))
db_country = tbl(dbcon, Id("survey_data", "country"))

db_cmr = tbl(dbcon, Id("survey_data", "cmr"))
db_taxa = tbl(dbcon, Id("survey_data", "taxonomy"))

```

## All Penn Bd data

```{r}
# load metadata for survey_data
mdc = tbl(dbcon, Id("public", "all_columns")) %>%
  filter(table_schema == "survey_data") %>%
  collect()

# filter then load capture data table
db_capture = tbl(dbcon, Id("survey_data", "capture")) %>%
  select(all_of(tbl_keys("capture", mdc)),
         species_capture,
         bd_swab_id,
         microbiome_swab_id,
         bacterial_swab_id,
         life_stage,
         svl_mm) %>%
  filter(!is.na(bd_swab_id))
  
capture_chain = tbl_chain("capture", mdc, until="region")

db_capture_chain = tbl_join(dbcon, capture_chain, tbl=db_capture) %>%
  filter(date >= "2022-01-01",
         date < "2023-01-01",
         !is.na(site)) # drop orphan swab data currently missing site

db_capture_chain_bd = tbl(dbcon, Id("survey_data", "qpcr_bd_results")) %>%
  inner_join(db_capture_chain, by="bd_swab_id") %>%
  collect()

colnames(db_penn_bd)

write_csv(db_penn_bd, here("staging", "bd_query_01.csv"))

```

```{r}
bd_results = tbl(dbcon, Id("survey_data", "qpcr_bd_results")) %>%
  group_by(bd_swab_id) %>%
  mutate(count = n()) %>%
  ungroup() %>%
  filter(count > 1) %>%
  arrange(desc(count)) %>%
  collect()
  
```

# 1 - Import data

```{r}
data_capture = db_capture %>%
  left_join(db_survey, by = "survey_id") %>%
  left_join(db_visit, by = "visit_id" ) %>%
  left_join(db_site, by = "site_id") %>%
  left_join(db_region, by = "region_id") %>%
  left_join(db_country, by = "country_id") %>%
  mutate(year = year(date)) %>%
  filter(country == "brazil") %>%
  group_by(region, year) %>%
  summarise(count = n()) %>%
  collect() %>%
  arrange(year, region) %>%
  select(year, region, count)

data_ves = db_ves %>%
  left_join(db_survey, by = "survey_id") %>%
  left_join(db_visit, by = "visit_id" ) %>%
  left_join(db_site, by = "site_id") %>%
  left_join(db_region, by = "region_id") %>%
  left_join(db_country, by = "country_id") %>%
  mutate(year = year(date)) %>%
  filter(country == "brazil") %>%
  group_by(region, year) %>%
  summarise(count = n()) %>%
  collect() %>%
  arrange(year, region) %>%
  select(year, region, count)


obs_ves = db_ves %>%
  left_join(db_survey, by = "survey_id") %>%
  left_join(db_visit, by = "visit_id" ) %>%
  left_join(db_site, by = "site_id") %>%
  left_join(db_region, by = "region_id") %>%
  left_join(db_country, by = "country_id") %>%
  mutate(year = year(date)) %>%
  filter(country == "brazil") %>%
  collect()

b22cap = db_capture %>%
  left_join(db_survey, by = "survey_id") %>%
  left_join(db_visit, by = "visit_id" ) %>%
  left_join(db_site, by = "site_id") %>%
  left_join(db_region, by = "region_id") %>%
  left_join(db_country, by = "country_id") %>%
  mutate(year = year(date)) %>%
  filter(region == "boraceia",
         year == "2022") %>%
  collect() %>%
  arrange(taxon_capture, date)

b22ves = db_ves %>%
  left_join(db_survey, by = "survey_id") %>%
  left_join(db_visit, by = "visit_id" ) %>%
  left_join(db_site, by = "site_id") %>%
  left_join(db_region, by = "region_id") %>%
  left_join(db_country, by = "country_id") %>%
  mutate(year = year(date)) %>%
  filter(region == "boraceia",
         year == "2022") %>%
  collect() %>%
  arrange(taxon_ves, date)

all22cap = db_capture %>%
  left_join(db_survey, by = "survey_id") %>%
  left_join(db_visit, by = "visit_id" ) %>%
  left_join(db_site, by = "site_id") %>%
  left_join(db_region, by = "region_id") %>%
  left_join(db_country, by = "country_id") %>%
  mutate(year = year(date)) %>%
  filter(year == "2022",
         country == "brazil") %>%
  collect() %>%
  group_by(site, date) %>%
  summarise(count = n())

```